File: example/java/jimmer-sql-graphql/src/main/java/org/babyfish/jimmer/sql/example/graphql/entities/Chapter.java
package org.babyfish.jimmer.sql.example.graphql.entities;

import org.babyfish.jimmer.sql.*;
import org.babyfish.jimmer.sql.example.graphql.entities.common.BaseEntity;
import org.jetbrains.annotations.Nullable;

@Entity
public interface Chapter extends BaseEntity {

    @Id
    long id();

    /*
     * In database, the foreign key `book_id `of `chapter` cannot not null.
     *
     * However, `Chapter.book` is declared as nullable property,
     * because the parent object `Book` may be filtered by global filter
     * (the demo shows how to use global filter).
     */
    @Key
    @Nullable
    @ManyToOne(inputNotNull = true)
    @OnDissociate(DissociateAction.DELETE)
    Book book();

    @Key
    @Column(name = "chapter_no")
    int index();

    String title();
}

File: example/kotlin/jimmer-sql-graphql-kt/src/main/kotlin/org/babyfish/jimmer/example/kt/graphql/entities/Chapter.kt
package org.babyfish.jimmer.example.kt.graphql.entities

import org.babyfish.jimmer.example.kt.graphql.entities.common.BaseEntity
import org.babyfish.jimmer.sql.*

@Entity
interface Chapter : BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long

    @Key
    @ManyToOne(inputNotNull = true)
    @OnDissociate(DissociateAction.DELETE)
    val book: Book?

    @Key
    @Column(name = "chapter_no")
    val index: Int

    val title: String
}